HPCC-20639 Fix core in EsdlBindingImpl::qualifyServiceName()

afishbeck opened the pull request:
Don't need to set the method description and help this way as it can
be looked up as needed.  Also this method is called far to often to
cache the fairly static values every time.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeck commented:
@richardkchapman if there is an RC5, can we take this DESDL fix? If not, it can be moved to 7.02.

HPCCSmoketest commented:
This PR (11863) is in the 4/6 position of the queue. Estimated start time is after ~2.0 hour(s)

HPCCSmoketest commented:
This PR (11863) is in the 3/5 position of the queue. Estimated start time is after ~1.5 hour(s)

HPCCSmoketest commented:
This PR (11863) is in the 2/4 position of the queue. Estimated start time is after ~1.0 hour(s)

HPCCSmoketest commented:
This PR (11863) is in the 1/3 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
Process of PR-11863, label: desdlCore1 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 726f5dd575bed4a85e28234c4cd39b7074c6f548
Build: success
Install hpccsystems-platform-community_7.0.0-rc3.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  190 sec (00:03:10) |   94 sec (00:01:34) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1512 sec (00:25:12) |   19 sec (00:00:19) | 1912 sec (00:31:52) |

afishbeck commented:
@timothyklemm @gaojx please review
@gaojx we should get a build made of this and my fix coming later today so you can test.

timothyklemm commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -2274,6 +2274,45 @@ int EsdlBindingImpl::getQualifiedNames(IEspContext& ctx, MethodInfoArray & metho
+
+    if (!m_esdl)
+    {
+        ESPLOG(LogMax,"EsdlBindingImpl::getMethodDescription - ESDL definition for service not loaded");
```
Should this be getMethodProperty instead of getMethodDescription? Same question applies to other ESPLOG statements.

afishbeck replied:
Yes.  I refactored my fix without updating the log text.  Good catch.

timothyklemm commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -263,6 +263,11 @@ class EsdlBindingImpl : public CHttpSoapBinding
+    int getMethodProperty(IEspContext &context, const char *serv, const char *method, StringBuffer &page, const char *propname, const char *dfault);
+
+    virtual int getMethodDescription(IEspContext &context, const char *serv, const char *method, StringBuffer &page) override;
+    virtual int getMethodHelp(IEspContext &context, const char *serv, const char *method, StringBuffer &page) override;
```
The commit message states the description and help "can be looked up as needed. Does the absence of any calls to these new methods mean they aren't needed, or should they be called from somewhere?

afishbeck replied:
These are virtual overrides that are already called by ESP when needed to display these bits of text on forms etc.

afishbeck replied:
Yes.  I refactored my fix without updating the log text.  Good catch.

timothyklemm reviewed (changes requested):
Commit message includes two grammatical errors:
1. "they can be looked up" instead of "it can be looked up".
2. "far too often" instead of "far to often"

afishbeck committed:
HPCC-20639 Fix core in EsdlBindingImpl::qualifyServiceName()

Don't need to cache the method description and help this way as they
can be looked up as needed.  Also this method is called far too often
to write these fairly static values every time.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@timothyklemm @gaojx updated please review

HPCCSmoketest commented:
This PR (11863) is in the 1/1 position of the queue. Estimated start time is after ~0.5 hour(s)

gaojx commented:
Reviewed.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6bb2e47b98ee0f22bb5ed62b5dd76153f8fc1600
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |804 |2 |
|test (thor) |730 |717 |13 |
|test (roxie) |879 |866 |13 |


[Errors:](https://gist.github.com/e10dcd9ec83a4c21e8d6d4dfeab11c0c)
- test
  - hthor
    - 236\. [filecompcopy](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-filecompcopy-w20181024-123822-7-log), 487\. [persist_replicate](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-persist_replicate-w20181024-123942-2-log), 
  - thor
    - 221\. [filecompcopy](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-filecompcopy-w20181024-124300-log), 468\. [persist_replicate](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-persist_replicate-w20181024-124613-log), 585\. [spray_expire_test](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_expire_test-w20181024-124654-1-log), 586\. [spray_header_test (isTerminated=false)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_header_test-isterminated-false-w20181024-124654-8-log), 587\. [spray_header_test (isTerminated=true)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_header_test-isterminated-true-w20181024-124654-6-log), 588\. [spray_replicate_test](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_replicate_test-w20181024-124655-log), 590\. [spray_test (sprayFixed=false, sprayEmpty=false)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_test-sprayfixed-false,-sprayempty-false-w20181024-124655-6-log), 591\. [spray_test (sprayFixed=false, sprayEmpty=true)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_test-sprayfixed-false,-sprayempty-true-w20181024-124655-5-log), 589\. [spray_test (sprayFixed=true)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_test-sprayfixed-true-w20181024-124655-4-log), 594\. [spray_test_json (isSmallFile=false)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_test_json-issmallfile-false-w20181024-124656-2-log), 592\. [spray_test_json (isSmallFile=true, isUnBallanced=false)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_test_json-issmallfile-true,-isunballanced-false-w20181024-124655-7-log), 593\. [spray_test_json (isSmallFile=true, isUnBallanced=true)](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_test_json-issmallfile-true,-isunballanced-true-w20181024-124656-log), 657\. [superkey1](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-superkey1-w20181024-124717-5-log), 
  - roxie
    - 161\. [despray](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-despray-w20181024-125022-log), 510\. [persist_replicate](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-persist_replicate-w20181024-125242-1-log), 663\. [spray_dir_test](https://gist.github.com/HPCCSmoketest/e10dcd9ec83a4c21e8d6d4dfeab11c0c#file-spray_dir_test-w20181024-125334-log), 664\. [spray_
 ... ( comment is too long, 45 lines, 5953 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  185 sec (00:03:05) |    0 sec (00:00:00) |    2 sec (00:00:02) |   17 sec (00:00:17) | 1326 sec (00:22:06) |   19 sec (00:00:19) | 1565 sec (00:26:05) |

```

AttilaVamos commented:
@afishbeck Those errors related to my changes/experiments with Smoketest. I will reschedule this PR soon.

HPCCSmoketest commented:
Process of PR-11863, label: desdlCore1 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6bb2e47b98ee0f22bb5ed62b5dd76153f8fc1600
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |729 |1 |
|test (roxie) |879 |879 |0 |


[Errors:](https://gist.github.com/ce4dff23901ada04e1c949188355e0ff)
- test
  - thor
    - 657\. [superkey1](https://gist.github.com/HPCCSmoketest/ce4dff23901ada04e1c949188355e0ff#file-superkey1-w20181024-161616-5-log), 

HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  180 sec (00:03:00) |    0 sec (00:00:00) |    5 sec (00:00:05) |   18 sec (00:00:18) | 1301 sec (00:21:41) |   19 sec (00:00:19) | 1536 sec (00:25:36) |

richardkchapman commented:
Should this target 7.0.2?
